plugins/anx-udev/fu-anx-udev-device.c

```diff
@@ -0,0 +1,475 @@
+
+#include <string.h>
+
+//#include "fu-chunk.h"
+
+#include "fu-anx-udev-common.h"
+#include "fu-anx-udev-device.h"
+#include "fu-anx-udev-firmware.h"
+
+
+
+struct _FuAnxUdevDevice {
+	FuUsbDevice  parent_instance;
+	guint8       iface_idx;         /* bInterfaceNumber */
+	guint8       ep_num;                /* bEndpointAddress */
+	guint16      chunk_len;         /* wMaxPacketSize */
+	guint16      vid;
+	guint16      pid;
+	guint16      rev;
+};
+
+G_DEFINE_TYPE (FuAnxUdevDevice, fu_anx_udev_device, FU_TYPE_USB_DEVICE)
+static gboolean
+fu_anx_udev_device_send (FuAnxUdevDevice *self,
+                       AnxBbRqtCode reqcode,
+                       guint16 val0code,
+               		   guint16 index,
+                       guint8 *in,
+                       gsize in_len,
+                       GError **error)
+{
+    gsize actual_len = 0;
+	/* check size */
+	if (in_len > 64) {
+		g_set_error (error,
+					G_IO_ERROR,
+					G_IO_ERROR_INVALID_DATA,
+					"input buffer too large");
+		return FALSE;
+	}
+    g_return_val_if_fail (in != NULL, FALSE);
+	/* send data to device */
+	if (!g_usb_device_control_transfer (
+			fu_usb_device_get_dev (FU_USB_DEVICE (self)),
+			G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+			G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+			G_USB_DEVICE_RECIPIENT_DEVICE,
+			reqcode, /* request */
+			val0code, /* value */
+			index, /* index */
+			in, /* data */
+			in_len, /* length */
+			&actual_len, /* actual length */
+			(guint)ANX_BB_TRANSACTION_TIMEOUT,
+			NULL, error)) {
+		g_prefix_error (error, "send data error: ");
+		return FALSE;
+	}
+     if (actual_len != in_len)
+     {
+		g_prefix_error (error, "send data error count: ");
+         return FALSE;
+     }
+	return TRUE;
+}
+
+static gboolean
+fu_anx_udev_device_receive (FuAnxUdevDevice *self,
+               				AnxBbRqtCode reqcode,
+                       		guint16 val0code,
+               				guint16 index,
+                       		guint8 *out,
+                       		gsize out_len,
+                       		GError **error)
+{
+    gsize actual_len = 0;
+	/* check size */
+	if (out_len > 64) {
+		g_set_error (error,
+					G_IO_ERROR,
+					G_IO_ERROR_INVALID_DATA,
+					"output buffer too large");
+		return FALSE;
+	}
+    g_return_val_if_fail (out != NULL, FALSE);
+    /* get data from device */
+    if (!g_usb_device_control_transfer (
+					fu_usb_device_get_dev (FU_USB_DEVICE (self)),
+					G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					G_USB_DEVICE_RECIPIENT_DEVICE,
+					reqcode, /* request */
+					val0code, /* value */
+					index, /* index */
+					out, /* data */
+					out_len,  /* length */
+					&actual_len, /* actual length */
+					(guint)ANX_BB_TRANSACTION_TIMEOUT,
+					NULL, error)) {
+			g_prefix_error (error, "receive data error: ");
+			return FALSE;
+    }
+     if (actual_len != out_len)
+     {
+		g_prefix_error (error, "receive data error count: ");
+         return FALSE;
+     }

```

Comment:
Just to explain GError a bit; you use `g_prefix_error` to add detail to an error that's been set by something else, and then you use `g_set_error()` to set your own error. You need to do the latter here.